<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Цели-исследования" data-toc-modified-id="Цели-исследования-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Цели исследования</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Исследуйте-таблицы-—-выведите-первые-строки,-посчитайте-количество-строк-в-каждой-таблице" data-toc-modified-id="Исследуйте-таблицы-—-выведите-первые-строки,-посчитайте-количество-строк-в-каждой-таблице-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Исследуйте таблицы — выведите первые строки, посчитайте количество строк в каждой таблице</a></span><ul class="toc-item"><li><span><a href="#books" data-toc-modified-id="books-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>books</a></span></li><li><span><a href="#authors" data-toc-modified-id="authors-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>authors</a></span></li><li><span><a href="#publishers" data-toc-modified-id="publishers-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>publishers</a></span></li><li><span><a href="#ratings" data-toc-modified-id="ratings-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>ratings</a></span></li><li><span><a href="#reviews" data-toc-modified-id="reviews-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>reviews</a></span></li></ul></li><li><span><a href="#Задания" data-toc-modified-id="Задания-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Задания</a></span><ul class="toc-item"><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года;" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года;-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года;</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку;" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку;-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку;</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры;" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры;-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками;" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками;-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок." data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок.-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.</a></span></li></ul></li></ul></div>

## Загрузка данных

In [1]:
# импортируем библиотеки
import pandas as pd
import sqlalchemy as sa

In [2]:
# устанавливаем параметры
db_config = {
'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db' # название базы данных
}
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

In [3]:
# сохраняем коннектор
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})

In [4]:
# чтобы выполнить SQL-запрос, пишем функцию с использованием Pandas
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
    with engine.connect() as con:
        return pd.read_sql(sql=sa.text(query), con = con)

In [5]:
# формируем запрос и выводим данные
query = '''SELECT * FROM books LIMIT 5'''
get_sql_data(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

## Цели исследования

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.  
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.  
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.  

## Описание данных

**Таблица books**  
Содержит данные о книгах:  
book_id — идентификатор книги;  
author_id — идентификатор автора;  
title — название книги;  
num_pages — количество страниц;  
publication_date — дата публикации книги;  
publisher_id — идентификатор издателя.  
**Таблица authors**  
Содержит данные об авторах:  
author_id — идентификатор автора;  
author — имя автора.  
**Таблица publishers**  
Содержит данные об издательствах:  
publisher_id — идентификатор издательства;  
publisher — название издательства;  
**Таблица ratings**  
Содержит данные о пользовательских оценках книг:  
rating_id — идентификатор оценки;  
book_id — идентификатор книги;  
username — имя пользователя, оставившего оценку;  
rating — оценка книги.  
**Таблица reviews**  
Содержит данные о пользовательских обзорах:  
review_id — идентификатор обзора;  
book_id — идентификатор книги;  
username — имя автора обзора;  
text — текст обзора.  

## Исследуйте таблицы — выведите первые строки, посчитайте количество строк в каждой таблице

### books

In [6]:
query = '''SELECT * 
           FROM books 
           LIMIT 3
           '''
get_sql_data(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135

In [7]:
query = '''SELECT COUNT(*)
           FROM books 
           '''
get_sql_data(query)

count
0   1000

### authors

In [8]:
query = '''SELECT * 
           FROM authors
           LIMIT 3
           '''
get_sql_data(query)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie

In [9]:
query = '''SELECT COUNT(*)
           FROM authors 
           '''
get_sql_data(query)

count
0    636

### publishers

In [10]:
query = '''SELECT * 
           FROM publishers
           LIMIT 3
           '''
get_sql_data(query)

publisher_id  publisher
0             1        Ace
1             2   Ace Book
2             3  Ace Books

In [11]:
query = '''SELECT COUNT(*)
           FROM publishers 
           '''
get_sql_data(query)

count
0    340

### ratings

In [12]:
query = '''SELECT * 
           FROM ratings
           LIMIT 3
           '''
get_sql_data(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5

In [13]:
query = '''SELECT COUNT(*)
           FROM ratings
           '''
get_sql_data(query)

count
0   6456

### reviews

In [14]:
query = '''SELECT * 
           FROM reviews
           LIMIT 3
           '''
get_sql_data(query)

review_id  book_id      username  \
0          1        1  brandtandrea   
1          2        1    ryanfranco   
2          3        2      lorichen   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...

In [15]:
query = '''SELECT COUNT(*)
           FROM reviews
           '''
get_sql_data(query)

count
0   2793

## Задания

### Посчитайте, сколько книг вышло после 1 января 2000 года;

In [16]:
query = '''SELECT COUNT(book_id) 
           FROM books
           WHERE publication_date::date > '2000-01-01'
           '''
get_sql_data(query)

count
0    819

После 1 января 2000 года вышло 819 книг.

### Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [17]:
query = '''SELECT title,
                  COUNT (DISTINCT review_id) AS count_review,
                  AVG(rating) AS avg_rating
            FROM books AS b 
            JOIN ratings AS rat ON b.book_id = rat.book_id 
            JOIN reviews AS rev ON b.book_id = rev.book_id
            GROUP BY b.book_id
            LIMIT 3
            '''
get_sql_data(query)

title  count_review  avg_rating
0                                       'Salem's Lot             2    3.666667
1                 1 000 Places to See Before You Die             1    2.500000
2  13 Little Blue Envelopes (Little Blue Envelope...             3    4.666667

Количество обзоров и средний рейтинг.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [18]:
query = '''SELECT p.publisher,
                  COUNT(b.book_id) AS count_books
           FROM publishers AS p
           JOIN books AS b ON p.publisher_id = b.publisher_id
           WHERE num_pages > 50
           GROUP BY p.publisher
           ORDER BY count_books DESC
           LIMIT 3
           '''
get_sql_data(query)

publisher  count_books
0             Penguin Books           42
1                   Vintage           31
2  Grand Central Publishing           25

Издательство которое выпустило наибольшее число книг толще 50 страниц: Penguin Books

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [19]:
query = '''SELECT author, 
                  AVG(rating) AS avg_rating
            FROM authors AS a 
            JOIN books AS b ON a.author_id = b.author_id
            JOIN ratings AS r ON b.book_id = r.book_id
            WHERE b.book_id IN (SELECT book_id
                                FROM ratings 
                                GROUP BY book_id
                                HAVING COUNT(rating_id) >= 50 
                                ORDER BY COUNT(rating_id) DESC) 
            GROUP BY author 
            ORDER BY avg_rating DESC
            LIMIT 3       
            '''
get_sql_data(query)

author  avg_rating
0         J.K. Rowling/Mary GrandPré    4.287097
1  Markus Zusak/Cao Xuân Việt Khương    4.264151
2                     J.R.R. Tolkien    4.246914

Автор с самой высокой средней оценкой книг: J.K. Rowling/Mary GrandPré	

In [20]:
# ячейка ревьюера:

query = '''

SELECT author, AVG(avg_rating) AS avg_rating
FROM (
    SELECT author, AVG(rating) AS avg_rating
    FROM authors AS a 
    JOIN books AS b ON a.author_id = b.author_id
    JOIN ratings AS r ON b.book_id = r.book_id
    WHERE b.book_id IN (
        SELECT book_id
        FROM ratings 
        GROUP BY book_id
        HAVING COUNT(rating_id) >= 50 
        ORDER BY COUNT(rating_id) DESC) 
    GROUP BY author, b.book_id
    ORDER BY avg_rating DESC
) AS t
GROUP BY author
ORDER BY avg_rating DESC

'''
get_sql_data(query)

author  avg_rating
0                          J.K. Rowling/Mary GrandPré    4.283844
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151
2                                      J.R.R. Tolkien    4.258446
3                                   Louisa May Alcott    4.192308
4                                        Rick Riordan    4.080645
5                                     William Golding    3.901408
6                                       J.D. Salinger    3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
9                                           Dan Brown    3.754540
10                                         Lois Lowry    3.750000
11            George Orwell/Boris Grabnar/Peter Škerl    3.729730
12                                    Stephenie Meyer    3.662500
13                                     John Steinbeck    3.622951

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [21]:
query = '''WITH table1 AS (SELECT username, 
                                  COUNT(review_id)
                           FROM reviews
                           WHERE username IN (SELECT username
                                              FROM ratings
                                              GROUP BY username
                                              HAVING COUNT(rating) > 48) 
                           GROUP BY username)
           SELECT AVG(count)
           FROM table1
           '''
get_sql_data(query)

avg
0  24.0

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок: 24